homeassistant/components/fitbit/sensor.py

```diff
@@ -396,264 +445,149 @@ class FitbitSensorEntityDescription(SensorEntityDescription):
-        ]
-        if "devices/battery" in monitored_resources:
-            devices = api.get_devices()
-            entities.extend(
-                [
-                    FitbitSensor(
-                        api,
-                        user_profile.encoded_id,
-                        config_path,
-                        FITBIT_RESOURCE_BATTERY,
-                        device,
-                    )
-                    for device in devices
-                ]
-            )
-        add_entities(entities, True)
-
-    else:
-        oauth = FitbitOauth2Client(
-            config_file.get(CONF_CLIENT_ID), config_file.get(CONF_CLIENT_SECRET)
+    config_file = await hass.async_add_executor_job(load_config_file, config_path)
+    _LOGGER.debug("loaded config file: %s", config_file)
+
+    if config_file is not None:
+        _LOGGER.debug("Importing existing fitbit.conf application credentials")
+        await async_import_client_credential(
+            hass,
+            DOMAIN,
+            ClientCredential(
+                config_file[CONF_CLIENT_ID], config_file[CONF_CLIENT_SECRET]
+            ),
+            auth_domain=AUTH_IMPL_IMPORTED,
         )
-
-        redirect_uri = f"{get_url(hass, require_ssl=True)}{FITBIT_AUTH_CALLBACK_PATH}"
-
-        fitbit_auth_start_url, _ = oauth.authorize_token_url(
-            redirect_uri=redirect_uri,
-            scope=[
-                "activity",
-                "heartrate",
-                "nutrition",
-                "profile",
-                "settings",
-                "sleep",
-                "weight",
-            ],
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data={
+                    "auth_implementation": AUTH_IMPL_IMPORTED,
+                    CONF_TOKEN: {
+                        ATTR_ACCESS_TOKEN: config_file[ATTR_ACCESS_TOKEN],
+                        ATTR_REFRESH_TOKEN: config_file[ATTR_REFRESH_TOKEN],
+                        "expires_at": config_file[ATTR_LAST_SAVED_AT],
+                    },
+                    CONF_CLOCK_FORMAT: config[CONF_CLOCK_FORMAT],
+                    CONF_UNIT_SYSTEM: config[CONF_UNIT_SYSTEM],
+                    CONF_MONITORED_RESOURCES: config[CONF_MONITORED_RESOURCES],
+                },
+            ),
         )
+    try:
+        await hass.async_add_executor_job(os.unlink, config_path)
+    except FileNotFoundError:
+        _LOGGER.debug("Unable to unlink fitbit configuration file")
 
-        hass.http.register_redirect(FITBIT_AUTH_START, fitbit_auth_start_url)
-        hass.http.register_view(FitbitAuthCallbackView(config, add_entities, oauth))
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
```

Comment:
One thing to note here, ADR-0021 is here now, and it requires us to let the user know what fails in a config flow. For example if the configuration was already imported we should let the user know. checkout `MELCloud` for an example. Also: This deprecated_yaml error message has been standardized in the home assistant domain, checkout LastFM for an example of using that standardized message

Reply:
w.r.t the config flow, so something like: failed to lookup your profile: therefore delete fitbit.conf and start over following the instructions on the website? Thanks!

Reply:
Euh, ADR21 is pretty vague in that regards, the approach I always take is just think of everything that can go wrong with importing and create a separate PR for everyone of them

Reply:
Will take a pass on additional config flow tests.

Reply:
I ended up keeping the custom text since there are some nuance with respect to `fitbit.conf` as an additional thing that needs to be cleaned up

However, i took from here the larger point about having friendlier error messages for the different cases, and improved them.

I plan to improve error handling even further with better error messages in follow up PRs to support things like re-auth and that will also be another change to improve import error messages as well to handle more cases.